### PR TITLE
feat: Support cookie authentication

### PR DESCRIPTION
**Pull request by khvn26**
Thanks for submitting a PR! Please check the boxes below:

- [ ] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

This adds a basic JWT cookie support. As long as `COOKIE_AUTH_ENABLED` setting is set to `True`: 
- The HTTP-only cookie is returned from registration and login endpoints. It then can be used to authenticate requests to endpoints that expect a token.
- The classic token is not being returned from registration and login endpoints. Login endpoint produces a 204 response.
- The CORS setting defaults are modified to allow credentials and the FE domain origin. All of the CORS settings are still settable individually for special cases / local development.

The JWT lifetime is set to be 10 hours by default. No refresh endpoints have been implemented at this point; when we do implement refreshing, token lifetime should be shortened significantly.

It's possible to invalidate the cookie via a `POST /api/v1/auth/logout/` request. The token is added to blacklist and the cookie is unset.

A signal handler is added to set `Access-Control-Allow-Origin` based on `get_current_site_url` utility function. The function is modified to properly use the Django sites framework cache.

## How did you test this code?

Added an integration test.

BE (including SAML) + FE has been tested in staging.

### CODE REVIEW

File: api/custom_auth/jwt_cookie/serializers.py
Comment (matthewelwell) [inline api/custom_auth/jwt_cookie/serializers.py:1]: Why do we need this empty file?

File: api/custom_auth/jwt_cookie/views.py
Comment (matthewelwell) [inline api/custom_auth/jwt_cookie/views.py]: I'm interested to know when this endpoint should be called? It feels like this should just be done on the generic logout endpoint?

File: api/custom_auth/urls.py
Comment (matthewelwell) [inline api/custom_auth/urls.py]: Hmm... shouldn't we only add this if the setting is enabled?

File: api/custom_auth/views.py
Comment (matthewelwell) [inline api/custom_auth/views.py]: Right, I was typing out some questions, because I didn't fully understand the behaviour here (and in the similar conditional statement above), but I think I understand it more now, although I still have some questions. 

As I understand it, we're probably adding some headers to the response so that the client has the JWT? What I'm still not sure about then is why we're returning an empty `Response` object, instead of doing something like: 

```python
response = self._action(serializer)
if settings.AUTH_JWT_COOKIE_ENABLED:
    return authorise_response(user, response)
return response
```

Comment (khvn26) [reply api/custom_auth/views.py]:
```
response = self._action(serializer)
```

creates the classic token and returns it. We don't need that for the cookie auth.

File: api/custom_auth/jwt_cookie/views.py
Comment (khvn26) [reply api/custom_auth/jwt_cookie/views.py]: >  the generic logout endpoint?

Which is...?

Comment (khvn26) [reply api/custom_auth/jwt_cookie/views.py]: I guess the real question here is why am I adding support for classic token deletion on logout. The answer is "no idea, it shouldn't be there."

I'm modifying it to only invalidate the JWT if it's present.

File: api/custom_auth/jwt_cookie/serializers.py
Comment (khvn26) [reply api/custom_auth/jwt_cookie/serializers.py:1]: Good catch, removed.

File: api/custom_auth/urls.py
Comment (khvn26) [reply api/custom_auth/urls.py]: Good point, done.

File: api/custom_auth/jwt_cookie/views.py
Comment (khvn26) [reply api/custom_auth/jwt_cookie/views.py]: This is done.

File: api/custom_auth/urls.py
Comment (khvn26) [reply api/custom_auth/urls.py]: Reversed this to simplify testing. The endpoint is a no-op for classic token auth anyway now.

File: api/custom_auth/views.py
Comment (matthewelwell) [reply api/custom_auth/views.py]: I guess this means that the API changes depending on an environment variable, which is slightly odd? But maybe it's ok.

Comment (matthewelwell) [reply api/custom_auth/views.py]: I guess we have similar cases elsewhere so yeah, this is fine.

File: api/custom_auth/urls.py
Comment (matthewelwell) [inline api/custom_auth/urls.py]: So, I guess what you're saying in other conversations is that `TokenDestroyView` does nothing, is that right? I guess I'm concerned about a couple of things here: 

 1. Are we removing functionality here for regular tokens? I assume not, but just checking. 
 2. Shouldn't we just have a `Logout` view class where we can internalise the conditional logic depending on the setting?

File: api/custom_auth/jwt_cookie/views.py
Comment (matthewelwell) [reply api/custom_auth/jwt_cookie/views.py]: Sorry, I think I misunderstood this code before. I didn't really notice that we were subclassing the `TokenDestroyView`. I guess this now relates to the question that I've asked [here](https://github.com/Flagsmith/flagsmith/pull/4662#discussion_r1777540740) in my second review.

Comment (khvn26) [reply api/custom_auth/jwt_cookie/views.py]: Yeah I got misled by your comment as well. Will revert tomorrow.

File: api/custom_auth/urls.py
Comment (khvn26) [reply api/custom_auth/urls.py]: `TokenDestroyView`, in fact, does remove the token. However, the endpoint has never been used by the frontend; to log the user out, the application simply removed the `t` cookie.

Now that I expect the endpoint to be actually used, I ultimately decided to remove function related to classic tokens and only implement the JWT logout.

I have renamed the view and added a docstring to hopefully be more clear.

File: api/custom_auth/jwt_cookie/views.py
Comment (khvn26) [reply api/custom_auth/jwt_cookie/views.py]: Responded [here](https://github.com/Flagsmith/flagsmith/pull/4662#discussion_r1777697979).

**Comment by rolodato** [comment]
Nit: I don't like the name `AUTH_JWT_COOKIE_ENABLED` - the fact that this cookie is a JWT is an implementation detail not relevant to any customers. We could have implemented an equivalent solution that uses opaque cookie tokens and not JWTs. In the context of OAuth/OIDC, the format of access tokens is only relevant to resource servers, i.e. Flagsmith API. Access tokens are meant to be opaque to clients (i.e. API-consuming applications, including the Flagsmith frontend).

It's likely that many customers will need to opt-in to this option if we decide to migrate everyone, so I'd like to have a useful name here. Nothing as permanent as a temporary solution etc.

How about `COOKIE_AUTH`, `COOKIE_AUTH_ENABLED` or similar instead?

**Comment by khvn26** [comment]
@rolodato Good point, changing the name to `COOKIE_AUTH_ENABLED`.

@kyle-ssg FYI

File: api/custom_auth/urls.py
Comment (matthewelwell) [reply api/custom_auth/urls.py]: Maybe the FE never used this functionality, but that's not to say that it's never been used by someone integrating with the API. It's obviously very unlikely, but not totally out of the question since we only provided API access via these user tokens for a long time. 

I don't see the harm in either: 

 1. Reverting to what you had initially and subclassing `TokenDestroyView` and calling `super()`
 2. Having 2 separate views that we add into urlpatterns depending on the existence of the environment variable.

Comment (khvn26) [reply api/custom_auth/urls.py]: > but that's not to say that it's never been used by someone integrating with the API

They will still have the `DELETE /api/v1/auth/token/` endpoint that serves this exact function, and arguably has better semantics as we do not delete classic tokens on logout.

I can add the notion to OpenAPI docs.

File: api/Makefile
Comment (khvn26) [inline api/Makefile]: Remove this once [this PR](https://github.com/Flagsmith/flagsmith-saml/pull/47) is merged.

**Comment by kyle-ssg** [comment]
In order to test this I had to add the following to common.py when testing against localhost:8000 from localhost:8080

```
# CORS_ORIGIN_ALLOW_ALL = True - Remove
CORS_ALLOW_CREDENTIALS = True
CORS_ALLOWED_ORIGINS = [
    "http://localhost:8080",
]
``` 

In addition, if we were to have this run in production, I believe the cookie domain would have to be ".flagsmith.com".

Do we need to make all of these things configurable? I don't think we can have ``CORS_ORIGIN_ALLOW_ALL = True`` if we run with this?

File: api/custom_auth/urls.py
Comment (kyle-ssg) [inline api/custom_auth/urls.py:29]: I think we need to delete the JWT_SLIDING_COOKIE_KEY cookie here, it looks to just invalidate rather than remove it too.

Comment (kyle-ssg) [reply api/custom_auth/urls.py:29]: I also noticed as a result, /auth/login/ wouldn't work until I force deleted the cookie. This makes me wonder if the backend needs to ignore the cookie on certain endpoints. 

I think this will cause an issue for invite workflows where an already "logged in" user registers / logs in with another account in order to accept an invite they created.

Comment (khvn26) [reply api/custom_auth/urls.py:29]: This is done.

**Comment by khvn26** [comment]
@kyle-ssg I have modified the CORS defaults so users won't have to set CORS settings manually for this.

File: api/custom_auth/urls.py
Comment (matthewelwell) [reply api/custom_auth/urls.py]: >They will still have the DELETE /api/v1/auth/token/ endpoint that serves this exact function, and arguably has better semantics as we do not delete classic tokens on logout.

But it's technically a breaking API change, right? We're removing functionality from an existing endpoint, that existed, and was documented for people to use.

File: Dockerfile
Comment (matthewelwell) [inline Dockerfile]: Remove this too before merge.

File: api/custom_auth/jwt_cookie/signals.py
Comment (matthewelwell) [inline api/custom_auth/jwt_cookie/signals.py:12]: Is there some documentation that we can link to for this?

File: api/custom_auth/urls.py
Comment (khvn26) [reply api/custom_auth/urls.py]: It seems you are correct here:

image

Still, I'd prefer to deprecate this functionality and direct people to `DELETE /api/v1/auth/token/` (as it will result in "logout" too).

Comment (matthewelwell) [reply api/custom_auth/urls.py]: We can't just remove it though. I don't mind adding a warning that it'll be deprecated in the next version (and let's add it to the V2 API issue).

File: frontend/common/stores/account-store.js
Comment (matthewelwell) [inline frontend/common/stores/account-store.js]: nit but there's an inconsistency in the formatting of some of these ternary statements.

```javascript
Project.cookieAuthEnabled?'true' : res.key
```

here vs the following in subsequent lines


```javascript
Project.cookieAuthEnabled ? 'true' : data.token
```

File: infrastructure/aws/staging/ecs-task-definition-web.json
Comment (matthewelwell) [inline infrastructure/aws/staging/ecs-task-definition-web.json]: Do we also need to remove this before merge @khvn26 ?

File: api/custom_auth/jwt_cookie/signals.py
Comment (khvn26) [reply api/custom_auth/jwt_cookie/signals.py:12]: Not really, you have to read django-cors-headers source for this knowledge.

Comment (matthewelwell) [reply api/custom_auth/jwt_cookie/signals.py:12]: Understood.

File: api/custom_auth/urls.py
Comment (khvn26) [reply api/custom_auth/urls.py]: The problem is that it's being called in FE now. We need to either add conditional logic to FE or explain to users that their tokens get invalidated if they logout explicitly. The latter is fine by me as using an admin API key is the clear alternative.

File: infrastructure/aws/staging/ecs-task-definition-web.json
Comment (khvn26) [reply infrastructure/aws/staging/ecs-task-definition-web.json]: Yes.

File: frontend/common/stores/account-store.js
Comment (khvn26) [reply frontend/common/stores/account-store.js]: Fixed.

File: api/custom_auth/urls.py
Comment (matthewelwell) [reply api/custom_auth/urls.py]: > their tokens get invalidated if they login explicitly

I assume here you meant to say 'logout explicitly' ? 

It doesn't seem like it'd be hard to add conditional logic to the frontend to only call the logout endpoint if the jwt cookie exists, and seems to be the right thing to do to maintain compatibility, but I can't speak for @kyle-ssg here. 

Alternatively, we do what you suggested, and just show a warning when the user hits the logout button to say that any direct API integrations they have using the user token will stop functioning.

Another alternative (that doesn't involve any more FE work) is still to conditionally change the view class used by the `logout` route depending on the existence of environment variable, but we may then need to think about whether we can actually run both in parallel.

Comment (khvn26) [reply api/custom_auth/urls.py]: As discussed offline, the FE conditional logic was added and token deletion function restored in the BE.

File: frontend/common/stores/account-store.js
Comment (matthewelwell) [inline frontend/common/stores/account-store.js:335]: ... is the semi-colon here some horrible JS syntax that I'm not aware of, or a typo? @kyle-ssg

**Review by matthewelwell — APPROVED** [review]
Approving with one minor comment

**Pull request merged**
